code/datums/components/clothing_dirt.dm

```diff
@@ -0,0 +1,90 @@
+/// This component applies tint to clothing when its exposed to pepperspray, used in /obj/item/clothing/mask/gas.
+
+/datum/component/clothing_dirt
+	/// The ITEM_SLOT_* slot the item is equipped on, if it is.
+	var/equipped_slot
+
+	/// Mob wearing the clothing
+	var/mob/living/carbon/wearer
+
+	/// Amount of dirt stacks on the clothing
+	var/dirtiness = 0
+
+	/// Clothing we're applying tint on
+	var/obj/item/clothing/clothing
+
+/datum/component/clothing_dirt/Initialize()
+	if(!isclothing(parent))
+		return COMPONENT_INCOMPATIBLE
+
+/datum/component/clothing_dirt/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ATOM_EXAMINE, PROC_REF(on_examine))
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, PROC_REF(on_equip))
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, PROC_REF(on_drop))
+	RegisterSignal(parent, COMSIG_COMPONENT_CLEAN_ACT, PROC_REF(on_clean))
+	RegisterSignal(parent, COMSIG_ATOM_EXPOSE_REAGENTS, PROC_REF(on_expose))
+	clothing = parent
+
+/datum/component/clothing_dirt/UnregisterFromParent()
+	clothing.tint -= dirtiness
+	clothing = null
+	if (!isnull(wearer))
+		wearer.update_tint()
+		UnregisterSignal(wearer, COMSIG_ATOM_EXPOSE_REAGENTS)
+		wearer = null
+	else
+		UnregisterSignal(parent, COMSIG_ATOM_EXPOSE_REAGENTS)
+	UnregisterSignal(parent, list(
+		COMSIG_ATOM_EXAMINE,
+		COMSIG_ITEM_EQUIPPED,
+		COMSIG_MOB_UNEQUIPPED_ITEM,
+		COMSIG_COMPONENT_CLEAN_ACT,
+	))
+	return ..()
+
+/datum/component/clothing_dirt/proc/on_equip(datum/source, mob/user, slot)
+	SIGNAL_HANDLER
+	if (!(slot & (ITEM_SLOT_MASK | ITEM_SLOT_HEAD)))
+		return
+	UnregisterSignal(parent, COMSIG_ATOM_EXPOSE_REAGENTS)
+	wearer = user
```

Comment:
We don't need to track wearer in a datum var. 

Registering the signal would pass the wearer as the first argument of the signal proc. So you can just get it from there.

Reply:
For the wash proc latter down the line, you can just do `clothing.loc` to find the wearer of the parent.

Reply:
i think its better to actually have a pointer to the wearer because its simpler to check for it that way. it saves the trouble of checking if the mask is actually equipped in the slot or not in on_drop proc and the unregister proc and is just more readable imho.

Reply:
The problem is the hanging reference can be problematic for, say, hard deletes. Something we want to avoid, especially when it's not necessary, like here

Reply:
```dm
/datum/component/clothing_dirt/proc/on_drop(obj/item/source, mob/user)
	SIGNAL_HANDLER
	UnregisterSignal(user, COMSIG_ATOM_EXPOSE_REAGENTS) // does nothing if it's not registered
	RegisterSignal(parent, COMSIG_ATOM_EXPOSE_REAGENTS, PROC_REF(on_expose), override = TRUE) // set to override to do nothing if it's already registered
```
```dm
/datum/component/clothing_dirt/proc/on_expose(atom/target, list/reagents, datum/reagents/source, methods)
	SIGNAL_HANDLER

	if(iscarbon(target) && !is_protected(target))
		return
```
```dm
/datum/component/clothing_dirt/proc/is_protected(mob/living/carbon/wearer)
	return wearer.head && (wearer.head.flags_cover & PEPPERPROOF)
```
```dm
/datum/component/clothing_dirt/proc/on_clean(datum/target, clean_types)
	SIGNAL_HANDLER
	var/obj/item/clothing/clothing = parent
	if (clean_types & (CLEAN_WASH|CLEAN_SCRUB))
		clothing.tint -= dirtiness
		dirtiness = 0
		if(iscarbon(clothing.loc))
			var/mob/living/carbon/wearer = clothing.loc
			wearer.update_tint()
```

Reply:
alright, i didnt realise that if ther signal was not found on the target the unregister proc doesnt do anything, so yea wearer is gone